core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -159,13 +166,14 @@ static void prepareHeaders(Metadata headers, DecompressorRegistry decompressorRe
   @Override
   public void start(final Listener<RespT> observer, Metadata headers) {
     checkState(stream == null, "Already started");
-    checkNotNull(observer, "observer");
-    checkNotNull(headers, "headers");
+    this.observer = checkNotNull(observer, "observer");
+    this.headers = checkNotNull(headers, "headers");
 
     if (context.isCancelled()) {
       // Context is already cancelled so no need to create a real stream, just notify the observer
       // of cancellation via callback on the executor
       stream = NoopClientStream.INSTANCE;
+      streamStarted = true;
```

Comment:
I removed the streamStarted here. It was unnecessary.  However, I found it's hard to get rid of the streamStarted, mostly because stream.cancel(...), stream.request(...) cannot be called before stream.start(..) is called, so there has to be a way to track that.